darshan-util/darshan-logutils.c

```diff
@@ -2069,7 +2069,7 @@ static int darshan_log_get_namerecs_3_00(void *name_rec_buf, int buf_len,
  * Gets list of modules present in logs and returns the info
  */
 void darshan_log_get_modules(darshan_fd fd,
-                             struct darshan_mod_info **mods,
+                             struct darshan_mod_info mods[DARSHAN_MAX_MODS],
```

Comment:
I'm not sure what putting a static sized array in the definition does here actually?  C doesn't know how to do bounds checks.  It might be more conventional to just match the logutils.h function prototype (with `struct darshan_mod_info *mods`).

More generally, this API doesn't do anything that obligates the caller to pass in a sufficiently large array.  We could make the `count` argument an input/output argument as one option (on the way in it is the size of the array, on the way out it is how many array elements were populated).  Or make another function to report how many modules are present, since it won't change once the log exists.

Reply:
Actually I retract the last sentence above.  Even if there were a separate helper fn to inform the caller of the module count, this function still can't know if the caller made the array big enough :)  Maybe just make the count an in/out and call it a day.

Reply:
It just decays to a pointer yeah, although in 2 dimensions I believe the compiler will actually barf on misuse. Can't you use an assert to enforce the static size if you don't trust the caller? It should always be `DARSHAN_MAX_MODS` rather than negotiable, no?

Reply:
By convention the caller should make it DARSHAN_MAX_MODS for sure.  I don't think there is any technical way to enforce or assert that it's that big, though.  I could be proven wrong :)

I think the more C-like strategy is for the caller to tell the function how big the buffer is rather than making its size implied.  The former is a little more explicit/pedantic from an API perspective and makes sure that (for example) you don't somehow accidentally get a mismatch on what the value of DARSHAN_MAX_MODS is between the function and the caller.

Reply:
arrghh, `sizeof` is going to resolve to the size of a single struct b/c you feed in a pointer... fat pointers like Rust/C++ would be nice. You could make your own "smart type" (augment the `struct` I suppose) by using a termination sentinel and iterating over the input array I guess. C isn't memory safe, so I'm pretty sure a sufficiently creative person can cause trouble with pointers into the abyss anyway.

Or maybe even mark/name a new function for CFFI usage only. Either approach seems better than the leak in production. Feel free to push in whatever changes you like.